src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -367,16 +358,18 @@ def _get_uniquely_named_nec_pnads(
     def get_name(cls) -> str:
         return "backchaining"
 
-    def _spawn_new_pnad(self, pnad: PartialNSRTAndDatastore,
-                        segment: Segment) -> PartialNSRTAndDatastore:
-        """Given a general PNAD and some segment with necessary add effects
-        that the PNAD must achieve, create a new PNAD ("spawn" from the most
-        general one) so that it has the necessary add effects contained in the
-        given segment.
+    def _spawn_new_pnad(self, segment: Segment) -> PartialNSRTAndDatastore:
+        """Given a PNAD and some segment with necessary add effects that the
+        PNAD must achieve, create a new PNAD ("spawn" from the most general one
+        associated with the segment's option) so that it has the necessary add
+        effects contained in the given segment.
 
         Returns the newly constructed PNAD, without modifying the
         original.
         """
+        # Create a general PNAD for the segment's option.
+        pnad = self._create_general_pnad_for_option(
+            segment.get_option().parent)
         # Assert that the segment contains necessary_add_effects.
         necessary_add_effects = segment.necessary_add_effects
         assert necessary_add_effects is not None
```

Comment:
move "# Assert that this really is a general PNAD." and the subsequent line a few lines up, to right after we create the pnad